misc/keycloak-test-helper/src/main/java/org/keycloak/test/TestsHelper.java

```diff
@@ -38,207 +40,307 @@
+            return clientRepresentation;
+        }
 
-    public static String initialAccessCode;
+        public String getRegistrationCode() {
+            return registrationCode;
+        }
+    }
 
-    public static String appName;
-    
-    public static int initialAccessTokenCount = 2; 
+    public static final String DEFAULT_KEYCLOAK_URL = "http://localhost:8080/auth";
+    public static final String DEFAULT_ADMIN_USERNAME = "admin";
+    public static final String DEFAULT_ADMIN_PASSWORD = "admin";
+    public static final String DEFAULT_ADMIN_REALM = "master";
+    public static final String DEFAULT_ADMIN_CLIENT = "admin-cli";
+    public static final String DEFAULT_TEST_REALM = DEFAULT_ADMIN_REALM;
+    public static final String DEFAULT_USER_ROLE = "user";
+
+    protected final String keycloakBaseUrl;
+    protected final String adminUserName;
+    protected final String adminPassword;
+    protected final String adminClient;
+    protected final String adminRealm;
+
+    protected String testRealm;
+    protected Keycloak keycloak;
+    protected String accessToken;
+    protected volatile boolean isInitialized;
+    protected Map<String, ClientData> createdClients = new HashMap<>();
+
+    /**
+     * Creates a new helper instance.
+     */
+    public TestsHelper() {
+        this(DEFAULT_ADMIN_USERNAME, DEFAULT_ADMIN_PASSWORD);
+    }
 
-    protected static String clientConfiguration;
+    /**
+     * Creates a new helper instance.
+     *
+     * @param adminUserName Admin username.
+     * @param adminPassword Admin password.
+     */
+    public TestsHelper(String adminUserName, String adminPassword) {
+        this(DEFAULT_KEYCLOAK_URL, adminUserName, adminPassword, DEFAULT_ADMIN_REALM, DEFAULT_ADMIN_CLIENT, DEFAULT_TEST_REALM);
+    }
 
-    protected static String registrationAccessCode;
+    /**
+     * Creates a new helper instance.
+     *
+     * @param keycloakBaseUrl Full keycloak URL.
+     * @param adminUserName Admin username.
+     * @param adminPassword Admin password.
+     * @param adminRealm Master realm name.
+     * @param adminClient Admin Client name.
+     * @param testRealm new instance.
+     */
+    public TestsHelper(String keycloakBaseUrl, String adminUserName, String adminPassword, String adminRealm, String adminClient, String testRealm) {
+        this.keycloakBaseUrl = keycloakBaseUrl;
+        this.testRealm = testRealm;
+        this.adminUserName = adminUserName;
+        this.adminPassword = adminPassword;
+        this.adminRealm = adminRealm;
+        this.adminClient = adminClient;
+    }
 
-    public static String createClient(ClientRepresentation clientRepresentation) {
-        ClientRegistration reg = ClientRegistration.create()
-                .url(keycloakBaseUrl, testRealm)
-                .build();
+    /**
+     * Initialization method.
+     *
+     * @return <code>this</code>
+     */
+    public TestsHelper init() {
```

Comment:
Is there any specific reason behind this `init` method or we could just initialize things in the constructor. This would avoid having to call `init` when using this API too ?

Reply:
Yeah, we could do it in the constructor but calling non-final method in the constructor is general an antipattern (I think even Josh wrote on it in Effective Java but my old memory might not be accurate here). And making those methods final would make this whole class less flexible (as you probably noticed, there are plenty of protected methods and fields to encourage inheritance there). 

So when I summed all the above up, I though the best approach is to use a good old `#init()` method :)

Reply:
I would avoid exposing too much public/protected methods and I still don't see a reason behind that `init` method. But you know the requirements better (especially those behind encouraging inheritance), so ... LGTM.